wandb/sdk_py27/wandb_monitor.py

```diff
@@ -0,0 +1,313 @@
+                    "to_table returned an incompatible object: {}".format(table)
+                )
+
+        self._flush_count += 1
+        self._maybe_rotate_run()
+        wandb.log(metrics)
+
+    # TODO: make byte size factor into our buffer?
+    def _call_size_bytes(self):
+        total = 0
+        if self._schema:
+            for key in ["inputs", "outputs"]:
+                for arg in self._schema[key]:
+                    total += arg.bytes
+        return total
+
+    @property
+    def estimated_buffer_bytes(self):
+        return self._call_size_bytes() * self._max_samples
+
+    def _data_type(self, obj, source=None, key=None):
+        # TODO: handle sequences / tensors
+        if wandb.util.is_numpy_array(obj):
+            return ArgType(key, source, "np", obj.shape, obj.nbytes)
+        elif wandb.util.is_pandas_data_frame(obj):
+            return ArgType(key, source, "df", obj.shape, obj.to_numpy().nbytes)
+        else:
+            return ArgType(key, source, None, None, sys.getsizeof(obj))
+
+    def _detect_schema(self, result, args, kwargs):
+        self._schema = {"inputs": []}
+        for key, obj in enumerate(args):
+            self._schema["inputs"].append(self._data_type(obj, "args", key))
+        for key, obj in kwargs.items():
+            self._schema["inputs"].append(self._data_type(obj, "kwargs", key))
+        self._schema["outputs"] = [self._data_type(result, "results", 0)]
+        estimated_bytes = self.estimated_buffer_bytes
+        if estimated_bytes > self.BUFFER_WARNING_BYTES:
+            wandb.termwarn(
+                "@wandb.monitor estimates {} of memory will be consumed.\nConsider reducing max_call_samples (currently {})".format(
+                    human_size(estimated_bytes), self._max_samples
+                )
+            )
+
+
+def monitor(
+    to_table = None,
+    name_or_artifact = None,
+    max_call_samples = 32,
+    flush_interval = 60,
+):
+    """
+    Function decorator for performantely monitoring predictions during inference.
+    You must call `wandb.init` before using this decorator.  It also requires that
+    numpy is available in your environment.
+
+    Attributes:
+        to_table (lambda): A function which returns a `wandb.Table` and accepts a
+            sampled list of `Call` objects.
+        name_or_artifact (str, Artifact): The name or Artifact instance to store
+            the data visualization table.
+        max_call_samples (int): The maximum number of calls to sample an buffer in
+            memory
+        flush_interval (int): The number of seconds to buffer calls before flushing
+            to W&B
+
+    Examples:
+        Basic usage
+        ```python
+        wandb.init(project="monitoring")
+
+        def to_table(calls):
+            table = wandb.Table("Input", "Output")
+            for call in calls:
+                table.add_data([wandb.Image(call.args[0]), np.argmax(call.results[0])])
+            return table
```

Comment:
We really need a type system that tells us what the underlying data is.  Just guessing by the shape will be really error prone, i.e.

- timeseries data (28,10)
- NLP embeddings (1000, 30)
- greyscale image (28,28)
- rgb image (28,28,3)
- rgba image (28,28,4)
- weird medical image (28,28,N)

It's also not clear if the predict function would receive raw user data, or preprocessed / normalized data.  Then we also need a primitive to handle softmax outputs and a Class mapping to output categories.  For bounding boxes and segmentation masks being able to guess what outputs represent what could be really tricky as well.

Also, we don't know if one of the dimensions actually represents a batch.  And it could be the first dimension or the last one...  Anyway, it's a rabbit hole.